Randomize app id on add-on load

machaboo opened the pull request:
So here is my proposed solution, we make all settings optional except minSeeders (which we default to 1, seems reasonable), and to no default app id at all (what's the point? it creates more issues then it solves)

If there is no app id set in the settings by the user, it then creates an id from 2 random words (added a script to get the names based on some module that does this). In my tests, i got these app_id's: open_support, least_lion, driving_therefore, paragraph_distance, explore_speed. Which seem more natural then just random characters.

I've tested and we can keep using `persist` for the token, as even if we change the app_id, the token stays the same and the api works fine.

machaboo committed:
Randomise app id on add-on load

machaboo committed:
Remove logs

machaboo committed:
Simplify config

machaboo committed:
Make minSeeders config required

machaboo commented:
btw, it is possible that the initial app id got shadow banned because of the hard coded token. tokens seem to be unique to each user, and they don't change based on app id, so sharing a token to multiple users might have been considered abuse, and thus the app id associated with it banned.

machaboo commented:
oh, i think we have a new problem, according to the docs:
https://torrentapi.org/apidocs_v2.txt?app_id=Jackett

> To use this api you will need to first to get yourself a token
> get_token=get_token
> Tokens automaticly expire in 15 minutes.
> The api has a 1req/2s limit.

I haven't yet seen my token expire, but we should somehow safeguard against possible expiration..

machaboo committed:
Consider token stale after 15 mins

machaboo committed:
Remove logs

machaboo committed:
2 second timeout between token and next request

machaboo commented:
Ok, I've done 2 more changes:

- we now consider tokens that have been fetched more then 15 mins ago stale, and refresh them

- I've added a 2 second timeout to the promise resolve of fetching the token, without this change, the first ever request always failed, it makes sense now after reading the docs, because we were doing the get token request and the next one in less then 2 seconds

machaboo commented:
maybe it would also be a good idea to persist the generated app id? what do you think?

sleeyax commented:
Woah, first of all thanks a lot for your contribution and dedication to this project :)

> I haven't yet seen my token expire, but we should somehow safeguard against possible expiration..

I haven't seen my token expire either, which is why I didn't implement any checks for this initially. But yeah, better save than sorry.

> 
> 
> maybe it would also be a good idea to persist the generated app id? what do you think?

You're right. Considering randomly generated app_ids work fine anyways, we can just remove it from user settings (bcs like you mentioned there isn't really a good point to leave it there, except for advanced users maybe).

machaboo commented:
I would still leave a configurable app id, but i would remove `minSeeders`, there are few seeded 4k torrents anyway..

machaboo commented:
i'd leave the app id because if we persist the randomly generated one, and it gets banned, users will have no way of fixing it, and i would also log some error if `response.statusCode == 429` (too many requests) as that is the most common error, and should be logged comprehensively, we could even recommend changing the app id and restarting the add-on if we get that status code

machaboo commented:
we could even automate "fixing" it by changing to a new app id and clearing the token if we get a 429 status code.. but that might be overengineering..

sleeyax commented:
> 
> 
> we could even automate "fixing" it by changing to a new app id and clearing the token if we get a 429 status code.. but that might be overengineering..

I was just about to comment this. You can probably check if the status code is anything other than `200`.

machaboo committed:
Persist auto-generated app id

machaboo committed:
Retry once on request failure

machaboo committed:
Small fix

machaboo committed:
Add comprehensive log when requests fail

machaboo committed:
Use 1 seed minimum for all requests

machaboo committed:
Remove minSeeders from config

machaboo commented:
More changes:

- we are now persisting the auto-generated app id

- i've added logic to retry once (with a different app id / token) if status code is different then `200` and there is no user defined app id

- added a comprehensive log when requests fail that also suggests changing the app id in settings and restarting the add-on

- removed `minSeeders` from config and set the minimum seeds to 1 for all requests

machaboo commented:
i think we're done.. although all this is hard to test, but the logic seems solid..

one more thing to think about is if we should set a 2 second timeout on the retrying logic (because of the 1 request per 2 second rule), i'm not sure we should though, as that limitation seems to be per token, as far as i understand it..

machaboo committed:
Add timeout of 2 seconds to retry logic

machaboo commented:
i've added a 2 sec timeout to the retry logic for safety, you can remove it if you don't agree with it..

should be all set now.. as long as the logic holds, i can't imagine what else could go wrong with this add-on..

machaboo committed:
Add default option for "Sort" config

sleeyax commented:
Code could be refactored/simplified a bit, but other than that it looks good to me. I'm merging it.

machaboo commented:
can u release it on github so it updates in PMS too? or do u plan to change something else before releasing?

sleeyax commented:
@machaboo oh right, I completely forgot. Thanks for letting me know

Edit: I just released it as v0.0.3

machaboo commented:
great, thanks!
